### ISSUE TITLE
App base path for multiple WASM hosted solution + CSS isolation + RCL support

### ISSUE THREAD

**Issue by guardrex** [dotnet/AspNetCore.Docs#25836]
https://github.com/dotnet/aspnetcore/issues/41625

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: be59faf1-5a20-b7e9-19ff-06589023bb15
* Version Independent ID: 250e7458-3c1d-0117-261d-c03ac136f16a
* Content: [Host and deploy ASP.NET Core Blazor WebAssembly](https://docs.microsoft.com/en-us/aspnet/core/blazor/host-and-deploy/webassembly?view=aspnetcore-6.0#static-assets-and-class-libraries-for-multiple-blazor-webassembly-apps)
* Content Source: [aspnetcore/blazor/host-and-deploy/webassembly.md](https://github.com/dotnet/AspNetCore.Docs/blob/main/aspnetcore/blazor/host-and-deploy/webassembly.md)
* Product: **aspnet-core**
* Technology: **aspnetcore-hostdeploy**
* GitHub Login: @guardrex
* Microsoft Alias: **riande**

**Comment by guardrex** [comment]
@philip-reed ... I'm very bogged down in high priority doc work for Blazor Hybrid ⛰️⛏️😅 at the moment, which releases shortly. I won't be able to get back to this until the middle of next week at the earliest, and that's if I don't get slammed with even more Hybrid work by then.

The code that you've referenced for hosting multiple client WASM apps works in the base case. The 😈 **_is in the details tho_**, as you can see. This section will never be able to lay out code and guidance for specific use cases for all of the different combos of technology and hosting. It's more of a general guide for the developer to then customize to their specific app and hosting specs. That code worked recently when I made updates to call out that the hosting ASP.NET Core app can also work like a normal RP/MVC app. That's what a reader was asking for. WRT the app base path, I'm not sure if the section can provide additional information ... it's going to depend on the root cause of the problem that you're facing. I'll need to investigate further using your exact use case to see what the topic can call out.

If you haven't already done so with your current repro project that you put up for the PU issue, I recommend that you make that app as close to your use case as possible. IIRC, you were combining an RCL with multiple hosted WASM apps and having trouble getting the apps to load, Razor components to render, and/or components to render with the right static assets (styling, etc.). When I reach this it will be helpful for me to have a repro with whatever you're trying to accomplish and your description of the exact foul behaviors that you're seeing with it. Then, I'll see if I can get it running and report back. If the topic can receive a useful update and/or cross-links to existing content, then I'll take it into a PR.

As it stands at the moment, I hope to reach this next week. It might take longer tho 🏃🏃🏃🏃. You can surely get help from the community if it's a pressing problem in the meantime. We recommend the usual places ...

* [Stack Overflow (tag: `blazor`)](https://stackoverflow.com/questions/tagged/blazor)
* [ASP.NET Core Slack Team](http://tattoocoder.com/aspnet-slack-sign-up/)
* [Blazor Gitter](https://gitter.im/aspnet/Blazor)

**Comment by philip-reed** [comment]
Thanks. I have the beginnings of a repro with hosting multiple apps.

Once I get this working the issue was the same in that there was a 404
while trying to get the imported css for the RCL.

Will update once this is done.

Thanks.

**Comment by guardrex** [comment]
@philip-reed ... Very good ... and the RCL angle is a new scenario for this coverage. In past discussions and updates for this multiple hosted WASM coverage, the focus was never on the use of RCLs in this scheme. We probably will get to the bottom of it and then be able to add a piece to the section for whatever we find out.

**Comment by guardrex** [comment]
**UPDATE (5/17)**: I'm still slammed with work on Blazor Hybrid for the upcoming release ... it's a **_TOP PRORITY_** 🚓🚑🚒, and that's not likely to change until next week at the earliest. In the meantime, the community might be able to assist on public channels until I can get back to this.

* [Stack Overflow (tag: `blazor`)](https://stackoverflow.com/questions/tagged/blazor)
* [ASP.NET Core Slack Team](http://tattoocoder.com/aspnet-slack-sign-up/)
* [Blazor Gitter](https://gitter.im/aspnet/Blazor)

**Comment by guardrex** [comment]
@philip-reed ... I had a minute this morning 🏃😅 to look.

Here's what I did ...

1. Start with the multiple hosted WASM apps guidance. I have an existing example that I was using for that other issue that I mentioned where the dev was merely trying to make the hosting ASP.NET Core app also serve Razor Pages. I'll just use that since it's sitting here ready to use.
1. I added a RCL to the solution. I then add a reference to the new RCL project in the first and second client apps.
1. I placed the RCL's `Component1` into each `Counter` component of the client apps (`<ComponentLibrary.Component1 />`).
1. Ran the host app (`MultipleBlazorApps.Server`) to see what the behavior is thus far. No styling and no import of the RCL's bundled styles in `https://localhost:5001/MultipleBlazorApps.SecondClient.styles.css` or `https://localhost:5002/MultipleBlazorApps.SecondClient.styles.css` ...

   ![Capture](https://user-images.githubusercontent.com/1622880/169285084-d00b6f3b-7a6b-4d29-bb19-1afabff201f9.PNG)

   ![Capture1](https://user-images.githubusercontent.com/1622880/169285097-b6a00404-0c87-4ccf-a62f-cd45d5aa7f71.PNG)

1. Manually add the bundled stylesheet asset to the `index.html` of each client app ...

   ```html
   <link href="_content/ComponentLibrary/ComponentLibrary.bundle.scp.css" rel="stylesheet" />
   ```

1. ... and that seems to have licked it 🎉 without any further changes to the guidance in the multiple hosted WASM coverage ...

   ![Capture2](https://user-images.githubusercontent.com/1622880/169286078-16c9f0a9-337c-43e2-8465-ee3b46fca1d8.PNG)

   ![Capture3](https://user-images.githubusercontent.com/1622880/169286091-db22d3ab-9600-49c9-9859-e1ebb030a5da.PNG)

..... so .....

You were correct to find and remark on the guidance at ...

https://docs.microsoft.com/en-us/aspnet/core/blazor/host-and-deploy/webassembly?view=aspnetcore-6.0#static-assets-and-class-libraries-for-multiple-blazor-webassembly-apps

... and it's clear to me that that's **_NOT CLEAR_** for this scenario with an RCL's bundled styles. It wasn't present merely because I probably assumed that it would work auto-magically, like when you reference an RCL from a regular Blazor project. I think an additional line and example makes sense for this scenario. I also need to see how cross-linking is set up in the multiple WASM section for that *Static assets and class libraries for multiple Blazor WebAssembly apps* section to make sure that it's easy to find the new line/example that I'll add.

Leave this issue open. It will close automatically when the PR merges later.

**Comment by philip-reed** [comment]
Yes! I'm glad you managed to recreate it.

And yes adding the css link in index.html does solve the issue.

However, the rcl css is still imported (with an incorrect url?) into each
client apps css, and this 404's for me. Do you get the same?

If the advice is to manually link the style sheet in index.html, how do we
stop the 404?

Hope that makes sense, I'm typing this from a phone while walking around a
warehouse.

**Comment by guardrex** [comment]
No ... didn't see that ... there's no such `@import` here and no 404s in my test app. I'm not sure what's different about your solution that produced that. You'll need to research that further on your own. I'm toooooo buried in work right now to look ⛰️⛏️😅. I'll have time in ........ uhhhhhhh ..... mmmmmmm 🤔 ....... **_in 2025!!!!!_** 🤣 

**_Note to self_**: 🤔 A paragraph with a list of steps to explain **_exactly what to do to add an RCL_** for the example multiple hosted WASM app probably makes sense ... a quick list of those steps ☝️ with the relevant `<link>` to add.

**Comment by guardrex** [comment]
**UPDATE**: Still trying to get back and wrap this up. It will take a little more time to reach it. 🏃⛰️⛏️😅
